src/Orchard.Web/Modules/Orchard.Azure/Services/FileSystems/AzureFileSystem.cs

```diff
@@ -260,6 +273,11 @@ public class AzureFileSystem {
             path = ConvertToRelativeUriPath(path);
             newPath = ConvertToRelativeUriPath(newPath);
 
+            if (FileSystemStorageProvider.FileNameContainsInvalidCharacters(Path.GetFileName(newPath), out var invalidCharacters)) {
+                throw new ArgumentException(
+                    string.Format("The new file name contains invalid character(s): {0}", string.Join(", ", invalidCharacters)));
```

Comment:
If that shows to the user, translate

Reply:
Addressed in 08ae5ddccf2772be2afb2d37fdb7ce6918df0865.
I created a new type of exception to be able to handle such cases correctly and without breaking changes or reaching through abstractions.

Reply:
The strings passed when building those exceptions are still un-translated in AzureFileSystem.cs (while they are localizable in other places). It seems like that message is then never used directly, because in the `catch` statements a different message is "forwarded" as error text: however, I think it would still be better to have these strings localizable, so we can use the exception message directly.

Reply:
That's what we discussed yesterday with Sebastien and after your comment I dug even deeper into it, but localization doesn't work for `AzureFileSystem`, because it's not a registered type in the DI container, which the `LocalizationModule` relies on to find where translations can be injected. I added an Autofac module to register `AzureFileSystem` and with that, `LocalizationModule` finds it, but translated strings aren't applied still. I'm not sure where the process gets stuck, but I don't think it's worth pursuing this further.

This is already a big improvement, because:
- This validation is now on the storage provider-level, so it doesn't just apply to Media Library.
- Validation is now applied for the Azure storage provider too with the same logic as the file system.
- Media Library didn't have validation at all possible places in the first places.
- `AzureFileSystem` already had non-localized exception messages, so unless the Azure storage is heavily refactored, we'll just have to live with that.